### PR TITLE
Add support for foreign functions.

### PR DESCRIPTION
**Pull request by zhenjunMa**
recently,  i want to call the host function in my wasm,  i found there is a spec define this interface `proxy_call_foreign_function `

more info: https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT#foreign-function-interface-ffi

my wasm is written by golang, so i use this repo in my project, but it doesn't implement `proxy_call_foreign_function `, so i do this.

### CODE REVIEW

**Comment by zhenjunMa** [comment]
this one  #163

File: proxywasm/hostcall.go [hunk 1]
```diff
@@ -82,6 +82,22 @@ func GetHttpCallResponseTrailers() (types.Trailers, error) {
 	return ret, types.StatusToError(st)
 }
 
+func DispatchForeignFuncCall(funcName string, param string) (ret string, err error) {
```
Comment (mathetake) [inline proxywasm/hostcall.go:85]: this is not async function but synchronous and nor a callout, so `Dispatch` is not an appropriate name.

And I definitely don't think you should expose the mangling scheme of `param` to users. Could you please add an example under `examples` directory and demonstrate how this is used?

Comment (mathetake) [inline proxywasm/hostcall.go:85]:
```suggestion
func CallForeignFunction(funcName string, param string) (ret string, err error) {
```

File: proxywasm/rawhostcall/rawhostcall.go [hunk 2]
```diff
@@ -80,6 +80,9 @@ func ProxyHttpCall(upstreamData *byte, upstreamSize int, headerData *byte, heade
 	bodyData *byte, bodySize int, trailersData *byte, trailersSize int, timeout uint32, calloutIDPtr *uint32,
 ) types.Status
 
+//export proxy_call_foreign_function
+func ProxyCallForeignFunction(funcNamePtr *byte, funcNameSize int, paramPtr *byte, paramSize int, returnData **byte, returnSize *int) types.Status
+
```
Comment (mathetake) [inline proxywasm/rawhostcall/rawhostcall.go:85]: could you add this to `rawhostcall_mock.go` as well so we can test this locally?

**Review by mathetake — CHANGES_REQUESTED** [review]

File: e2e/e2e_test.go [hunk 3]
```diff
@@ -104,11 +104,19 @@ func Test_E2E(t *testing.T) {
 		staticReply: 8010,
 		admin:       28310,
 	}, accessLogger))
+
```
Comment (mathetake) [inline e2e/e2e_test.go:107]: unnecessary change.

File: e2e/e2e_test.go [hunk 4]
```diff
@@ -104,11 +104,19 @@ func Test_E2E(t *testing.T) {
...
 	t.Run("dispatch_call_on_tick", testRunnerGetter(envoyPorts{
 		endpoint:    11011,
 		staticReply: 8011,
 		admin:       28311,
 	}, dispatchCallOnTick))
+
```
Comment (mathetake) [inline e2e/e2e_test.go:113]: remove the empty line to line with the style.

File: e2e/e2e_test.go [hunk 5]
```diff
@@ -104,11 +104,19 @@ func Test_E2E(t *testing.T) {
...
+	t.Run("foreign_call_on_tick", testRunnerGetter(envoyPorts{
+		endpoint:    11011,
+		staticReply: 8011,
+		admin:       28311,
+	}, callForeignOnTick))
+
```
Comment (mathetake) [inline e2e/e2e_test.go:119]: ditto.

File: e2e/e2e_test.go [hunk 6]
```diff
@@ -104,11 +104,19 @@ func Test_E2E(t *testing.T) {
...
 	t.Run("dispatch_call_on_tick", testRunnerGetter(envoyPorts{
 		endpoint:    11011,
 		staticReply: 8011,
 		admin:       28311,
 	}, dispatchCallOnTick))
+
+	t.Run("foreign_call_on_tick", testRunnerGetter(envoyPorts{
+		endpoint:    11011,
+		staticReply: 8011,
+		admin:       28311,
```
Comment (mathetake) [inline e2e/e2e_test.go:117]: use the different ports so we can avoid the port collision.

File: proxywasm/hostcall.go [hunk 7]
```diff
@@ -82,6 +82,22 @@ func GetHttpCallResponseTrailers() (types.Trailers, error) {
 	return ret, types.StatusToError(st)
 }
 
+func CallForeignFunction(funcName string, param string) (ret string, err error) {
```
Comment (mathetake) [inline proxywasm/hostcall.go:85]: let's have []byte as a type for param which is more generic 
```suggestion
func CallForeignFunction(funcName string, param []byte) (ret string, err error) {
```

File: proxytest/root.go [hunk 8]
```diff
@@ -263,6 +263,27 @@ func (r *rootHostEmulator) ProxyHttpCall(upstreamData *byte, upstreamSize int, h
 	return types.StatusOK
 }
 
+var funcMap = map[string]func(string)string {
```
Comment (mathetake) [inline proxytest/root.go:266]:
```suggestion
var foreignFunctions = map[string]func([]byte)[]byte {
```

**Review by mathetake — CHANGES_REQUESTED** [review]
Sorry, I was kind of confused so please ignore that comment (`And I definitely don't think you should expose the mangling scheme of param to users.`).

You can use "compress" function in Envoy https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/wasm/foreign.cc#L30 to pass e2e test.

File: proxywasm/hostcall.go (hunk 7 above)
Comment (mathetake) [reply proxywasm/hostcall.go:85]: and also have the return to be []byte as well
```suggestion
func CallForeignFunction(funcName string, param string) (ret []byte, err error) {
```

File: e2e/e2e_test.go [hunk 9]
```diff
@@ -98,7 +98,6 @@ func Test_E2E(t *testing.T) {
 		staticReply: 8009,
 		admin:       28309,
 	}, configurationFromRoot))
-
```
Comment (mathetake) [inline e2e/e2e_test.go:101]: thanks!

File: examples/foreign_call_on_tick/README.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+## call_foreign_function_on_tick
+
+this example periodically call foreign functions
+
+//TODO
```
Comment (mathetake) [inline examples/foreign_call_on_tick/README.md:5]: Could you resolve?

**Review by mathetake — COMMENTED** [review]
Thanks, almost there.

**Review by mathetake — APPROVED** [review]
Thanks @zhenjunMa

**Pull request merged**